### PR TITLE
Basic functionality to send tokens to other users directly

### PR DESCRIPTION
**Pull request by JasoonS**

### CODE REVIEW

File: matic/contracts/WildcardSteward_matic_v2.sol [hunk 1]
```diff
@@ -6,20 +6,12 @@ import "../vendered/@openzeppelin/contracts-ethereum-package-3.0.0/contracts/Ini
...
 import "./interfaces/IDai.sol";
 
-// import "./GSNRecipientBase.sol";
-
-// import "../vendered/gsn-2.0.0-beta.1.3/contracts/BaseRelayRecipient.sol";
-// import "../vendered/gsn-2.0.0-beta.1.3/contracts/interfaces/IKnowForwarderAddressGsn.sol";
-
-// import "@nomiclabs/buidler/console.sol";
-
-contract WildcardSteward_matic_v1 is Initializable, BasicMetaTransaction {
```
Comment (JasoonS) [inline matic/contracts/WildcardSteward_matic_v2.sol:22]: Just cleanup

File: matic/contracts/WildcardSteward_matic_v2.sol [hunk 2]
```diff
@@ -134,6 +126,14 @@ contract WildcardSteward_matic_v1 is Initializable, BasicMetaTransaction {
...
     event ChangeAuctionParameters();
 
+    event Transfer(
+        uint256 indexed tokenId,
+        address indexed from,
+        address indexed to,
+        uint256 price,
+        uint256 deposit
+    );
+
```
Comment (JasoonS) [inline matic/contracts/WildcardSteward_matic_v2.sol:136]: New event

File: matic/contracts/WildcardSteward_matic_v2.sol [hunk 3]
```diff
@@ -613,7 +613,7 @@ contract WildcardSteward_matic_v1 is Initializable, BasicMetaTransaction {
         internal
         returns (bool transferSuccess)
     {
-        return paymentToken.transferFrom(msgSender(), address(this), amount);
+        return paymentToken.transferFrom(from, address(this), amount);
```
Comment (JasoonS) [inline matic/contracts/WildcardSteward_matic_v2.sol:616]: This was a bug, but luckily I think we only used this function where `from` was the `msgSender`

File: matic/contracts/WildcardSteward_matic_v2.sol [hunk 4]
```diff
@@ -1244,4 +1244,50 @@ contract WildcardSteward_matic_v1 is Initializable, BasicMetaTransaction {
...
+            );
+        require(
+            timeLeftOfDeposit > 604800, /* seconds in a week */
+            "Deposit less than 1 week"
+        );
+
+        receiveErc20(tokensDeposit, msgSender());
+
+        // We won't charge users to use this function:
+        // _distributePurchaseProceeds(tokenId);
```
Comment (JasoonS) [inline matic/contracts/WildcardSteward_matic_v2.sol:1283]: I'm not charging users to transfer the tokens. We can add this another time

File: matic/migrations/8_launch_tokens copy.js [hunk 5]
```diff
@@ -0,0 +1,184 @@
+const WildcardSteward_v3 = artifacts.require("WildcardSteward_v3_matic");
```
Comment (JasoonS) [inline matic/migrations/8_launch_tokens copy.js:1]: Whoops, will remove this file...

File: matic/contracts/WildcardSteward_v3_matic.sol [hunk 6]
```diff
@@ -1,1226 +0,0 @@
-// NOTE : this is actually the v0 matic contract.
```
Comment (JasoonS) [inline matic/contracts/WildcardSteward_v3_matic.sol:1]: Deleting this, the file is now called `_matic_v0` (which is our v3 contracts from mainnet deployed to matic basically)

File: matic/contracts/WildcardSteward_matic_v2.sol [hunk 7]
```diff
@@ -1244,4 +1244,50 @@ contract WildcardSteward_matic_v1 is Initializable, BasicMetaTransaction {
...
+        uint256 tokenId,
+        uint256 newPrice,
+        uint256 tokensDeposit
+    )
+        public
+        collectPatronageAndSettleBenefactor(tokenId)
+        collectPatronagePatron(to)
+        priceGreaterThanZero(newPrice)
+        youCurrentlyAreNotInDefault(msgSender())
+        youCurrentlyAreNotInDefault(to) // Maybe this condition is too strict? Maybe it should check foreclosure time after recieving the token rather?
```
Comment (moose-code) [inline matic/contracts/WildcardSteward_matic_v2.sol:1259]: Its fine I like this condition.

File: matic/contracts/WildcardSteward_matic_v2.sol [hunk 8]
```diff
@@ -1244,4 +1244,50 @@ contract WildcardSteward_matic_v1 is Initializable, BasicMetaTransaction {
 
         assetToken.transferFrom(_currentOwner, address(this), tokenId);
     }
+
+    function transferWithDeposit(
+        address to,
+        uint256 tokenId,
+        uint256 newPrice,
+        uint256 tokensDeposit
+    )
```
Comment (moose-code) [inline matic/contracts/WildcardSteward_matic_v2.sol:1253]: Possible attack vector. Lets not worry but keep this in mind.  

Its a stretch, but you can send a token with the min deposit (of 1 week) at quite a high price to another user, and after 1 week this token may be quite a strain on all the users deposit causing them to foreclose

**Review by moose-code — APPROVED** [review]
Cool it looks good. See comments, nice

File: matic/contracts/WildcardSteward_matic_v2.sol [hunk 9]
```diff
@@ -1244,4 +1244,50 @@ contract WildcardSteward_matic_v1 is Initializable, BasicMetaTransaction {
...
+
+        receiveErc20(tokensDeposit, msgSender());
+
+        // We won't charge users to use this function:
+        // _distributePurchaseProceeds(tokenId);
+
+        // Set the pecentage that wildcards will recieve when this token is purchased at 15%
+        serviceProviderPercentages[tokenId] = 150000;
+
+        deposit[to] = deposit[to].add(tokensDeposit);
```
Comment (moose-code) [inline matic/contracts/WildcardSteward_matic_v2.sol:1288]: Looking closer now but are we subtracting deposit from the sender?

Comment (moose-code) [reply matic/contracts/WildcardSteward_matic_v2.sol:1288]: nevermind I see its coming in the form of erc20 being taken from the sender, not from the senders deposit

File: matic/contracts/WildcardSteward_matic_v2.sol [hunk 10]
```diff
@@ -1244,4 +1244,50 @@ contract WildcardSteward_matic_v1 is Initializable, BasicMetaTransaction {
...
+        uint256 timeLeftOfDeposit =
+            tokensDeposit.mul(yearTimePatronagDenominator).div(
+                newPrice.mul(patronageNumerator[tokenId])
+            );
+        require(
+            timeLeftOfDeposit > 604800, /* seconds in a week */
+            "Deposit less than 1 week"
+        );
+
+        receiveErc20(tokensDeposit, msgSender());
```
Comment (moose-code) [inline matic/contracts/WildcardSteward_matic_v2.sol:1280]: ``
```suggestion
        assert(receiveErc20(tokensDeposit, msgSender()));
```

Comment (moose-code) [reply matic/contracts/WildcardSteward_matic_v2.sol:1280]: Just for more explicitness sake

File: matic/contracts/WildcardSteward_matic_v2.sol (hunk 7 above)
Comment (moose-code) [reply matic/contracts/WildcardSteward_matic_v2.sol:1259]: If the new user is a brand new user, I am guessing it won't show them in default? Just checking

File: matic/contracts/WildcardSteward_matic_v2.sol (hunk 8 above)
Comment (moose-code) [reply matic/contracts/WildcardSteward_matic_v2.sol:1253]: Basically you are sending someone a debt obligation

Comment (moose-code) [reply matic/contracts/WildcardSteward_matic_v2.sol:1253]: I guess the new user can immediately withdraw the deposit they received?

Comment (JasoonS) [reply matic/contracts/WildcardSteward_matic_v2.sol:1253]: Yeah, exactly, bad that I'm only looking at this now :grapes:

Comment (JasoonS) [reply matic/contracts/WildcardSteward_matic_v2.sol:1253]: I think it is fine, if it is enough to drain the users deposit after the week, then it is a large contribution to wildlife conservation too!

**Pull request merged**